File: backend/util/date_data_generator.py
from datetime import datetime
from dateutil.relativedelta import relativedelta

def group_by_date(items):
    result = {}
    for item in items:
        start_date = item['start_date']
        frequency = item['frequency']
        duration = 365
        dates = generate_dates(start_date, frequency, duration)
        for date in dates:
            if date in result:
                result[date].append(item)
            else:
                result[date] = [item]
    return result

def generate_dates(start_date, frequency, duration):
    dates = []
    date = datetime.strptime(start_date, '%Y-%m-%d')
    while (date - datetime.strptime(start_date, '%Y-%m-%d')).days <= duration:
        dates.append(date.strftime('%Y-%m-%d'))
        if frequency == 0:
            date += relativedelta(days=1)
        elif frequency == 1:
            date += relativedelta(weeks=1)
        elif frequency == 2:
            date += relativedelta(weeks=2)
        elif frequency == 3:
            date += relativedelta(months=1)
        elif frequency == 4:
            date += relativedelta(months=3)
        elif frequency == 5:
            date += relativedelta(months=6)
        elif frequency == 6:
            date += relativedelta(years=1)
        else:
            return dates
    return dates

File: plan-wise-react-node/backend/util/date_data_generator.js
const moment = require('moment');

function group_by_date(items) {
    let result = {};
    for (let item of items) {
        let start_date = item['start_date'];
        let frequency = item['frequency'];
        let duration = 365;
        let dates = generate_dates(start_date, frequency, duration);
        for (let date of dates) {
            if (date in result) {
                result[date].push(item);
            } else {
                result[date] = [item];
            }
        }
    }
    return result;
}

function generate_dates(start_date, frequency, duration) {
    let dates = [];
    let date = moment(start_date, 'YYYY-MM-DD');
    while (moment(date).diff(moment(start_date, 'YYYY-MM-DD'), 'days') <= duration) {
        dates.push(date.format('YYYY-MM-DD'));
        if (frequency === 0) {
            date.add(1, 'days');
        } else if (frequency === 1) {
            date.add(1, 'weeks');
        } else if (frequency === 2) {
            date.add(2, 'weeks');
        } else if (frequency === 3) {
            date.add(1, 'months');
        } else if (frequency === 4) {
            date.add(3, 'months');
        } else if (frequency === 5) {
            date.add(6, 'months');
        } else if (frequency === 6) {
            date.add(1, 'years');
        } else {
            return dates;
        }
    }
    return dates;
}

module.exports = { group_by_date, generate_dates };
